### PR TITLE
add `OpInfo`s for `nn.functional.binary_cross_entropy(_with_logits)?`

### PR DESCRIPTION
**Pull request by pmeier**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65681
* #68419
* #69212
* #67079
* __->__ #67023
* #67022
* #69211

cc @mruberry @jbschlosser @walterddr @kshitij12345

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -14268,6 +14289,58 @@ def ref_pairwise_distance(input1, input2):
...
         ),
+    ),
+    OpInfo(
+        "nn.functional.binary_cross_entropy",
+        sample_inputs_func=sample_inputs_binary_cross_entropy,
+        dtypes=floating_types(),
+        dtypesIfCUDA=floating_types_and(torch.float16, torch.bfloat16),
+        supports_out=False,
+        decorators=(
+            DecorateInfo(
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:14300]: I guess this is technically unnecessary because this test is xfailed below?

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -7572,6 +7572,27 @@ def sample_inputs_pdist(op_info, device, dtype, requires_grad, **kwargs):
...
+        *[((S, S), dict(reduction=reduction, weight=make((S, S)))) for reduction in reductions],
+    ]
+
+    return [
+        SampleInput(
+            (make if logits else make_prob)(shape, requires_grad=requires_grad),
+            args=(make_prob(shape, requires_grad=requires_grad),),
+            kwargs=kwargs,
+        )
+        for shape, kwargs in shapes_and_kwargs
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:7593]: style nit: yoda speak

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -7572,6 +7572,27 @@ def sample_inputs_pdist(op_info, device, dtype, requires_grad, **kwargs):
         *[SampleInput(make_input((S, S)), kwargs=dict(p=p)) for p in (0.0, 1.0, 2.0, 10.0, float("inf"))],
     ]
 
+def sample_inputs_binary_cross_entropy(op_info, device, dtype, requires_grad, logits=False, **kwargs):
+    make = partial(make_tensor, device=device, dtype=dtype)
+    make_prob = partial(make, low=0, high=1)
+
+    reductions = ("mean", "sum", "none")
+
+    shapes_and_kwargs = [
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:7581]: does this test the `pos_weight` arg of BINARY_CROSS_ENTROPY_WITH_LOGITS?

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -7572,6 +7572,27 @@ def sample_inputs_pdist(op_info, device, dtype, requires_grad, **kwargs):
         *[SampleInput(make_input((S, S)), kwargs=dict(p=p)) for p in (0.0, 1.0, 2.0, 10.0, float("inf"))],
     ]
 
+def sample_inputs_binary_cross_entropy(op_info, device, dtype, requires_grad, logits=False, **kwargs):
+    make = partial(make_tensor, device=device, dtype=dtype)
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:7576]: nice handling of probabilities vs. logits

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (pmeier) [reply torch/testing/_internal/common_methods_invocations.py:14300]: Yes, but without the decorator, the test would only complain about insufficient tolerances, which hides the actual issue. Thus if the source of the error is fixed, we need the decorator.

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (pmeier) [reply torch/testing/_internal/common_methods_invocations.py:7581]: Nope, I completely missed that, because it is not available for the op without logits. Good catch!

**Review by mruberry — APPROVED** [review]

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -7581,6 +7581,32 @@ def sample_inputs_pdist(op_info, device, dtype, requires_grad, **kwargs):
...
+    make_prob = partial(make, low=0, high=1)
+
+    reductions = ("mean", "sum", "none")
+
+    shapes_and_kwargs = [
+        *[(shape, None) for shape in ((), (1,), (S,), (S, S), (S, S, S))],
+        *[((S, S), dict(reduction=reduction)) for reduction in reductions],
+        *[((S, S), dict(reduction=reduction, weight=make((S, S)))) for reduction in reductions],
```
Comment (zou3519) [inline torch/testing/_internal/common_methods_invocations.py:7593]: The pytorch documentation suggests that the weight must be 1D: https://pytorch.org/docs/stable/generated/torch.nn.BCELoss.html#torch.nn.BCELoss

Are the docs wrong?

Comment (jbschlosser) [reply torch/testing/_internal/common_methods_invocations.py:7593]: Looks like the docs may be wrong: #68780

Or we need to tighten up what is considered a valid shape for weight, but it seems like there are use cases outside of 1D weight of shape (N).

**Comment by nkaretnikov** [comment]
a gentle reminder that this is still open as my pr https://github.com/pytorch/pytorch/pull/68951 depends on this (i'll update mine once this lands)

**Comment by pmeier** [comment]
cc @mruberry @saketh-are

**Comment by pmeier** [comment]
We already duplicated work now :slightly_frowning_face: #75604 @drisspg

**Comment by pmeier** [comment]
Superseded by #75782.
